Assets/MixedRealityToolkit.SDK/Experimental/FollowSolver/Follow.cs

```diff
@@ -422,47 +476,108 @@ private bool AngularClamp(Vector3 refPosition, Quaternion refRotation, Vector3 c
+                        }
+                        else if (angle > minMaxAngle)
+                        {
+                            rotation = Quaternion.AngleAxis(minMaxAngle - angle, Vector3.up) * rotation;
+                            angularClamped = true;
+                        }
+                    }
+                    break;
 
-                float newAngle = stepAngle * numberOfSteps;
+                case AngularClampType.AngleStepping:
+                    {
+                        float stepAngle = 360f / tetherAngleSteps;
+                        int numberOfSteps = Mathf.RoundToInt(SolverHandler.TransformTarget.transform.eulerAngles.y / stepAngle);
 
-                rotation = Quaternion.Euler(rotation.eulerAngles.x, newAngle, rotation.eulerAngles.z);
-            }
-            else
-            {
-                // This is negated because Unity is left-handed
-                float angle = AngleBetweenVectorAndPlane(toTarget, refRight);
-                float minMaxAngle = MaxViewHorizontalDegrees * 0.5f;
+                        float newAngle = stepAngle * numberOfSteps;
 
-                if (angle < -minMaxAngle)
-                {
-                    rotation = Quaternion.AngleAxis(-minMaxAngle - angle, Vector3.up) * rotation;
-                    angularClamped = true;
-                }
-                else if (angle > minMaxAngle)
-                {
-                    rotation = Quaternion.AngleAxis(minMaxAngle - angle, Vector3.up) * rotation;
-                    angularClamped = true;
-                }
+                        rotation = Quaternion.Euler(rotation.eulerAngles.x, newAngle, rotation.eulerAngles.z);
+                    }
+                    break;
+
+                case AngularClampType.RendererBounds:
+                case AngularClampType.ColliderBounds:
+                    {
+                        var extentsXZ = new Vector3(extents.x, 0.0f, extents.z);
+                        var extentsXZMagnitude = extentsXZ.magnitude;
+                        Vector3 min = refRotation * new Vector3(-extentsXZMagnitude, 0.0f, currentDistance);
+                        Vector3 max = refRotation * new Vector3(extentsXZMagnitude, 0.0f, currentDistance);
+
+                        // These are negated because Unity is left-handed
+                        float minAngle = -AngleBetweenOnXZPlane(toTarget, min);
+                        float maxAngle = -AngleBetweenOnXZPlane(toTarget, max);
+
+                        if (minAngle > 0)
+                        {
+                            rotation = Quaternion.AngleAxis(minAngle, Vector3.up) * rotation;
+                            angularClamped = true;
+                        }
+                        else if (maxAngle < 0)
+                        {
+                            rotation = Quaternion.AngleAxis(maxAngle, Vector3.up) * rotation;
+                            angularClamped = true;
+                        }
```

Comment:
Again, we should measure the angle once, and avoid rotating and constructing two new vectors. We should also be using `AngleBetweenVectorAndPlane` as this fixes a bug where objects could get stuck behind you. See lines 529-542.

Reply:
Thank you for the quick review! I missed these new methods during my merge. I'll remove the old ones and integrate these new ones.